agent/taskresource/cgroup/cgroup.go

```diff
@@ -0,0 +1,212 @@
+// express or implied. See the License for the specific language governing
+// permissions and limitations under the License.
+
+package taskresource
+
+import (
+	"encoding/json"
+	"os"
+	"path/filepath"
+	"strconv"
+	"sync"
+	"time"
+
+	"github.com/aws/amazon-ecs-agent/agent/api"
+	"github.com/aws/amazon-ecs-agent/agent/resources/cgroup"
+	"github.com/aws/amazon-ecs-agent/agent/utils/ioutilwrapper"
+	"github.com/cihub/seelog"
+	"github.com/containerd/cgroups"
+	"github.com/pkg/errors"
+)
+
+const (
+	memorySubsystem         = "/memory"
+	memoryUseHierarchy      = "memory.use_hierarchy"
+	rootReadOnlyPermissions = os.FileMode(400)
+)
+
+var (
+	enableMemoryHierarchy = []byte(strconv.Itoa(1))
+)
+
+// CgroupResource represents Cgroup resource
+type CgroupResource struct {
+	control             cgroup.Control
+	CgroupRoot          string
+	CgroupMountPath     string
+	ioutil              ioutilwrapper.IOUtil
+	createdAt           time.Time
+	desiredStatusUnsafe CgroupStatus
+	knownStatusUnsafe   CgroupStatus
+	// lock is used for fields that are accessed and updated concurrently
+	lock sync.RWMutex
+}
+
+// NewCgroupResource is used to return an object that implements the Resource interface
+func NewCgroupResource(control cgroup.Control, cgroupRoot string, cgroupMountPath string) *CgroupResource {
+	return &CgroupResource{
+		control:         control,
+		CgroupRoot:      cgroupRoot,
+		CgroupMountPath: cgroupMountPath,
+		ioutil:          ioutilwrapper.NewIOUtil(),
+	}
+}
+
+// SetDesiredStatus safely sets the desired status of the resource
+func (c *CgroupResource) SetDesiredStatus(status CgroupStatus) {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+
+	c.desiredStatusUnsafe = status
+}
+
+// GetDesiredStatus safely returns the desired status of the task
+func (c *CgroupResource) GetDesiredStatus() CgroupStatus {
+	c.lock.RLock()
+	defer c.lock.RUnlock()
+
+	return c.desiredStatusUnsafe
+}
+
+// SetKnownStatus safely sets the currently known status of the resource
+func (c *CgroupResource) SetKnownStatus(status CgroupStatus) {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+
+	c.knownStatusUnsafe = status
+}
+
+// GetKnownStatus safely returns the currently known status of the task
+func (c *CgroupResource) GetKnownStatus() CgroupStatus {
+	c.lock.RLock()
+	defer c.lock.RUnlock()
+
+	return c.knownStatusUnsafe
+}
+
+// SetCreatedAt sets the timestamp for resource's creation time
+func (c *CgroupResource) SetCreatedAt(createdAt time.Time) {
+	if createdAt.IsZero() {
+		return
+	}
+	c.lock.Lock()
+	defer c.lock.Unlock()
+
+	c.createdAt = createdAt
+}
+
+// GetCreatedAt sets the timestamp for resource's creation time
+func (c *CgroupResource) GetCreatedAt() time.Time {
+	c.lock.RLock()
+	defer c.lock.RUnlock()
+
+	return c.createdAt
+}
+
+// Create creates cgroup root for the task
+func (c *CgroupResource) Create(task *api.Task) error {
+	err := c.setupTaskCgroup(task)
+	if err != nil {
+		seelog.Criticalf("Unable to setup platform resources: %v", err)
```

Comment:
It'd be even better if we log task arn here and everywhere. Also, you can use this unofficial log format being followed in the engine package: `"Cgroup resource [%s] Unable to ..: %v", task.Arn, err"`. Can you please modify all log lines here to conform to that?

Example: https://github.com/aws/amazon-ecs-agent/blob/cc9fcff8e6977895cfafc49e04bdb4df3945b55a/agent/engine/docker_task_engine.go#L328